File: docs/workouts/templates/ios/WorkoutCatalogViewModel.swift
import Foundation

// Reference scaffold only. ViewModel exposes UI state and depends on domain use cases only.

@MainActor
final class WorkoutCatalogViewModel: ObservableObject {
    enum State: Equatable {
        case idle
        case loading
        case loaded([Exercise])
        case empty
        case failed(String)
    }

    @Published private(set) var state: State = .idle
    @Published var searchText: String = ""
    @Published var selectedLevel: WorkoutLevel?

    private let listExercises: ListExercisesUseCase

    init(listExercises: ListExercisesUseCase) {
        self.listExercises = listExercises
    }

    func load() async {
        state = .loading
        do {
            let filter = WorkoutCatalogFilter(searchText: searchText, level: selectedLevel)
            let exercises = try await listExercises.execute(filter: filter)
            state = exercises.isEmpty ? .empty : .loaded(exercises)
        } catch {
            state = .failed("Workouts could not load. Try again.")
        }
    }
}

File: docs/workouts/templates/android/WorkoutCatalogViewModel.kt
package app.tribelog.workouts.template

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

// Reference scaffold only. ViewModel depends on domain use cases, not Firestore or Compose.

sealed interface WorkoutCatalogUiState {
    data object Idle : WorkoutCatalogUiState
    data object Loading : WorkoutCatalogUiState
    data class Loaded(val exercises: List<Exercise>) : WorkoutCatalogUiState
    data object Empty : WorkoutCatalogUiState
    data class Failed(val message: String) : WorkoutCatalogUiState
}

class WorkoutCatalogViewModel(
    private val listExercises: ListExercisesUseCase
) : ViewModel() {
    private val mutableState = MutableStateFlow<WorkoutCatalogUiState>(WorkoutCatalogUiState.Idle)
    val state: StateFlow<WorkoutCatalogUiState> = mutableState.asStateFlow()

    var searchText: String = ""
    var selectedLevel: WorkoutLevel? = null

    fun load() {
        viewModelScope.launch {
            mutableState.value = WorkoutCatalogUiState.Loading
            try {
                val exercises = listExercises.execute(
                    WorkoutCatalogFilter(searchText = searchText, level = selectedLevel)
                )
                mutableState.value = if (exercises.isEmpty()) {
                    WorkoutCatalogUiState.Empty
                } else {
                    WorkoutCatalogUiState.Loaded(exercises)
                }
            } catch (_: Throwable) {
                mutableState.value = WorkoutCatalogUiState.Failed("Workouts could not load. Try again.")
            }
        }
    }
}

